Vic2ToHoI4/Source/Mappers/Provinces/ProvinceMapperFactory.cpp

```diff
@@ -0,0 +1,138 @@
+#include "ParserHelpers.h"
+#include "ProvinceMapper.h"
+#include "VersionedMappings.h"
+#include <fstream>
+
+
+
+namespace
+{
+
+std::optional<int> getNextProvinceNumFromFile(std::ifstream& definitions)
+{
+	std::string line;
+	getline(definitions, line);
+
+	if (const auto pos = line.find_first_of(';'); pos != std::string::npos)
+	{
+		try
+		{
+			return stoi(line.substr(0, pos));
+		}
+		catch (...)
+		{
+			Log(LogLevel::Warning) << "Bad line in /map/definition.csv: " << line;
+		}
+
+		return 0;
+	}
+
+	return std::nullopt;
+}
+
+
+void verifyProvinceIsMapped(const Mappers::VersionedMappings& versionedMapper, int provNum)
+{
+	if (!versionedMapper.getHoI4ToVic2Mapping().contains(provNum))
+	{
+		Log(LogLevel::Warning) << "No mapping for HoI4 province " << provNum;
+	}
+}
+
+
+void checkAllHoI4ProvincesMapped(const Mappers::VersionedMappings& versionedMapper,
+	 const Configuration& theConfiguration)
+{
+	std::ifstream definitions(theConfiguration.getHoI4Path() + "/map/definition.csv");
+	if (!definitions.is_open())
+	{
+		throw std::runtime_error("Could not open " + theConfiguration.getHoI4Path() + "/map/definition.csv");
+	}
+
+	while (true)
+	{
+		auto provNum = getNextProvinceNumFromFile(definitions);
+		if (!provNum)
+		{
+			break;
+		}
+		if (*provNum == 0)
+		{
+			continue;
+		}
+
+		verifyProvinceIsMapped(versionedMapper, *provNum);
+	}
+
+	definitions.close();
+}
+
+} // namespace
+
+
+
+Mappers::ProvinceMapper::Factory::Factory(const Configuration& theConfiguration)
+{
+	registerRegex(R"(\d\.[\d]+\.\d)", [this, theConfiguration](const std::string& unused, std::istream& theStream) {
+		const auto thisVersionsMappings = versionedMappingsFactory.importVersionedMappings(theStream);
+		checkAllHoI4ProvincesMapped(*thisVersionsMappings, theConfiguration);
+		provinceMapper->HoI4ToVic2ProvinceMap = thisVersionsMappings->getHoI4ToVic2Mapping();
+		provinceMapper->Vic2ToHoI4ProvinceMap = thisVersionsMappings->getVic2ToHoI4Mapping();
+	});
+	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
+}
+
+
+std::unique_ptr<Mappers::ProvinceMapper> Mappers::ProvinceMapper::Factory::importProvinceMapper(
+	 const Configuration& theConfiguration)
+{
+	Log(LogLevel::Info) << "Parsing province mappings";
+	provinceMapper = std::make_unique<ProvinceMapper>();
+
+	auto mapped = false;
+	for (const auto& mod: theConfiguration.getVic2Mods())
+	{
+		if (commonItems::DoesFileExist("Configurables/" + mod.getName() + "_province_mappings.txt"))
+		{
+			parseFile("Configurables/" + mod.getName() + "_province_mappings.txt");
+			mapped = true;
+			break;
+		}
+	}
+	if (!mapped)
+	{
+		parseFile("Configurables/province_mappings.txt");
+	}
+
+	return std::move(provinceMapper);
+}
+
+
+std::vector<int> Mappers::ProvinceMapper::getVic2ToHoI4ProvinceMapping(const int Vic2Province) const
+{
+	const auto mapping = Vic2ToHoI4ProvinceMap.find(Vic2Province);
+	if (mapping == Vic2ToHoI4ProvinceMap.end())
+	{
+		return {};
```

Comment:
Now, this is where there is no functional difference between an unmapped link to a province (map only on vic2 or only on hoi4 side deliberately), and a nonexistent link (due to a new province being added in a patch).

I know you have "verify all hoi4 provinces are mapped" function but I'd still throw warning when trying to access a province (from either v2 or hoi4 side) that doesn't exist in the mappings. It'd help ping issues with map patches to both hoi4 and stuff like hfm.